pkg/webhook/webhooks.go

```diff
@@ -0,0 +1,280 @@
+	}
+	wh := &Webhook{
+		Server: &http.Server{
+			Addr:      fmt.Sprintf(":%v", p.Port),
+			TLSConfig: &tls.Config{Certificates: []tls.Certificate{pair}},
+		},
+		KubeClient: kubeClient,
+		Clientset:  openebsClient,
+	}
+	return wh, nil
+}
+
+func admissionRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	// skip special kubernetes system namespaces
+	for _, namespace := range ignoredList {
+		if metadata.Namespace == namespace {
+			glog.Infof("Skip validation for %v for it's in special namespace:%v", metadata.Name, metadata.Namespace)
+			return false
+		}
+	}
+
+	annotations := metadata.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+
+	// admissionWebhookAnnotationValidateKey is the annotation key which can be use
+	// explicitly forcefully to disable the validation policy(enable by default)
+	// if set as 'false'. for ex: 'admission-webhook-openebs.io/validate: "false"'
+	// by default return 'true' if nothing is specified
+
+	var required bool
+	switch strings.ToLower(annotations[string(admissionWebhookAnnotationValidateKey)]) {
+	default:
+		required = true
+	case "n", "no", "false", "off":
+		required = false
+	}
+	return required
+}
+
+func validationRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	required := admissionRequired(ignoredList, metadata)
+	glog.Infof("Validation policy for %v/%v: required:%v", metadata.Namespace, metadata.Name, required)
+	return required
+}
+
+// validate validates the persistentvolumeclaim(PVC) delete request
+func (wh *Webhook) validate(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	req := ar.Request
+	status := &v1beta1.AdmissionResponse{}
+
+	// validates only if requested operation is DELETE and request kind is PVC
+	if ar.Request.Operation != v1beta1.Delete && req.Kind.Kind != "PersistentVolumeClaim" {
+		status.Allowed = true
+		return status
+	}
+	glog.Infof("AdmissionReview for Kind=%v, Namespace=%v Name=%v UID=%v patchOperation=%v UserInfo=%v",
+		req.Kind, req.Namespace, req.Name, req.UID, req.Operation, req.UserInfo)
+
+	// TODO* use raw object once validation webhooks support DELETE request
+	// object as non nil value https://github.com/kubernetes/kubernetes/issues/66536
+	//var pvc corev1.PersistentVolumeClaim
+	//err := json.Unmarshal(req.Object.Raw, &pvc)
+	//if err != nil {
+	//	glog.Errorf("Could not unmarshal raw object: %v, %v", err, req.Object.Raw)
+	//	status.Allowed = false
+	//	status.Result = &metav1.Status{
+	//		Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+	//		Message: err.Error(),
+	//	}
+	//	return status
+	//}
+
+	// fetch the pvc specifications
+	pvc, err := wh.KubeClient.CoreV1().PersistentVolumeClaims(req.Namespace).Get(req.Name, metav1.GetOptions{})
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
```

Comment:
I am not sure if this is a bad request?
validation will run in a loop.
this error comes as part of execution of validation. So how is the error related to bad request.

Reply:
Not sure ! here we can have different type of errors.. but changing to --> `InternalServerError`